server/src/main/scala/org/http4s/server/middleware/authentication/package.scala

```diff
@@ -11,14 +11,14 @@ package object authentication {
       challenge: Kleisli[F, Request[F], Either[Challenge, AuthedRequest[F, A]]])(
       routes: AuthedService[A, F])(implicit F: Sync[F]): HttpRoutes[F] =
     Kleisli { req =>
-      challenge
-        .mapF(OptionT.liftF(_))
-        .run(req)
-        .flatMap {
-          case Right(authedRequest) =>
-            routes(authedRequest)
-          case Left(challenge) =>
-            OptionT.some(Response(Status.Unauthorized).putHeaders(`WWW-Authenticate`(challenge)))
+      OptionT[F, Response[F]] {
+        F.flatMap(challenge(req)) {
+          case Left(challenge) => F.pure(Some(unauthorized(challenge)))
+          case Right(authedRequest) => routes(authedRequest).value
         }
+      }
     }
+
+  def unauthorized[F[_]](challenge: Challenge): Response[F] =
```

Comment:
I would make this private so we don't have to preserve it across future refactoring.